### ISSUE TITLE
destroying a memtable may take a long time

### ISSUE THREAD

**Issue by glommer** [scylladb/scylla#1551]
For the sake of testing, I am running scylla with cache disabled. In those situations the memtable destructor can take more than 100ms to execute. When the cache is enabled, all partitions in the memtable are moved to the cache and the partitions set is empty.

However, when the cache is disabled, instrumenting the memtable destructor yields the following example output:

```
Disposed` 110783 partions after 108998 usec`
```

That indicates that the time to free a partition is not that bad and is around 1 usec. But because we need to loop through them all synchronously, that can be an issue.

This is not a blocker because we will by default always run with the cache enabled. But it would be great to have this fixed nevertheless.

**Comment by tzach** [comment]
Will this affect memtable flush with cache?
In particular, when none of the data in memtable can be merged to cache

**Comment by belliottsmith** [comment]
Is this for a default cassandra-stress workload? If so 1usec seems a bit lengthy (to me), given the partitions are very small.

Anyway, while it may be that LSA free is not the most expensive part, this brings to mind #1442, and I wonder if another optimisation isn't possible to minimise the costs of rebalancing the heap (or histogram, should that be introduced), namely only marking the blocks that need rebalancing on free, and placing them in a queue, to only rebalance them when the queue reaches a certain size, and notably disabling rebalancing for the duration of a batch-wise free (like here).

Given the visitation order for the memtable will be random, it is likely a lot of unnecessary rebalancing is going on, since all of the blocks for the memtable are likely to be near empty when the process is done, and near full when it begins.

Of course, incrementally freeing would limit your latency impact here, but reducing the total cost is no doubt also helpful.

**Comment by glommer** [comment]
@belliottsmith yes, that is the default cassandra-stress workload. I am only taking an average over all partitions, so maybe the common case is not so bad, and this is driven upwards by the worst case?

We need to investigate further.

**Comment by glommer** [comment]
BTW, I did mention in the issue that this will affect only cases in which we run with the cache disabled, but that is not true: we don't call update_cache() directly on streaming, which means that streaming memtables will be destroyed this way as well.

**Comment by glommer** [comment]
@tgrabiec I wonder if this is still an issue after you have fixed memtable's clear() function to gracefully move entries to the cache instead of at once ?

**Comment by tgrabiec** [comment]
After 3511bf4a81b36e50ec772938eafa063f93a3aa44, clearing is done incrementally and will not block the reactor for the whole duration. However, I think the fact that clearing is slow is still something which we should improve upon, because memory is not released until clearing is complete. 

Some of the ideas were mentioned in this issue. I also recently opened https://github.com/scylladb/scylla/issues/1920
